File: app/src/main/java/ds/photosight/ui/widget/VotesWidget.java
package ds.photosight.ui.widget;


import android.content.Context;
import android.os.AsyncTask;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AccelerateDecelerateInterpolator;
import android.view.animation.Interpolator;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import com.androidquery.AQuery;
import ds.photosight.R;
import ds.photosight.utils.Utils;

import java.util.List;

@SuppressWarnings("ALL")
public class VotesWidget extends LinearLayout {

	private static final float ANIMATION_DURATION = 1500;
	private static int awardIconSize;// = Utils.dp(16);

	private List<Integer> mRates;
	private List<String> mAvards;
	private AQuery aq;
	private TextView[] mRateViews = new TextView[5];
	private View mGreenBar;
	private View mRedBar;


	public VotesWidget(final Context context, final AttributeSet attrs) {
		super(context, attrs);
		aq = new AQuery(this);

	}


	public void setRates(final List<Integer> rates) {
		mRates = rates;
	}


	public void setAvards(final List<String> avards) {
		mAvards = avards;
	}


	public void runAnimations() {

		awardIconSize = getAwardIconSize(mAvards.size());
		mRateViews[0] = aq.id(R.id.aRate).getTextView();
		mRateViews[1] = aq.id(R.id.oRate).getTextView();
		mRateViews[2] = aq.id(R.id.tRate).getTextView();
		mRateViews[3] = aq.id(R.id.lRate).getTextView();
		mRateViews[4] = aq.id(R.id.dRate).getTextView();
		for (TextView t : mRateViews) { t.setText(""); }

		mGreenBar = aq.id(R.id.greenBar).getView();
		mRedBar = aq.id(R.id.redBar).getView();
		((ViewGroup) aq.id(R.id.labelsContainer).getView()).startLayoutAnimation();
		removeAvards();

		new AsyncTask<List<Integer>, Integer, Boolean>() {

			@Override
			protected Boolean doInBackground(final List<Integer>[] params) {
				Interpolator interpolator = new AccelerateDecelerateInterpolator();
				List<Integer> values = params[0];
				float rating = 1 / (float) (values.get(3) + values.get(4)) * values.get(3);

				Integer[] results = new Integer[values.size() + 1];
				long start = System.currentTimeMillis();
				boolean escape = false;
				while (!escape) {
					try {
						Thread.sleep(10);
					} catch (InterruptedException e) {
						e.printStackTrace();
					}

					final float f = (System.currentTimeMillis() - start) / ANIMATION_DURATION;
					if (f >= 1)
						escape = true;

					for (int i = 0; i < values.size(); i++) {
						results[i] = (int) (values.get(i) * interpolator.getInterpolation(f));
					}

					//L.v("publish progress");
					results[results.length - 1] = (int) (rating * interpolator.getInterpolation(f) * 1000);
					publishProgress(results);

				}

				for (int i = 0; i < values.size(); i++) {
					results[i] = (int) (values.get(i));
				}
				publishProgress(results);

				return true;
			}


			@Override
			protected void onProgressUpdate(final Integer[] values) {
				for (int i = 0; i < values.length - 1; i++) {
					mRateViews[i].setText(String.valueOf(values[i]));
				}

				int height = values[values.length - 1] * mRedBar.getHeight() / 1000;
				//L.v("h=" + height);
				mGreenBar.getLayoutParams().height = height;
				//mGreenBar.setMinimumHeight();
			}


			@Override
			protected void onPostExecute(final Boolean o) {
				if (mAvards != null)
					addAvards();
			}
		}.execute(mRates);
	}


	private int getAwardIconSize(final int size) {
		if (size > 7)
			return Utils.dp(16);
		else if (size > 4)
			return Utils.dp(24);
		else
			return Utils.dp(32);
	}


	private void removeAvards() {
		((ViewGroup) aq.id(R.id.avardsContainer).getView()).removeAllViews();
	}


	private void addAvards() {
		final ViewGroup vg = (ViewGroup) aq.id(R.id.avardsContainer).getView();

		LinearLayout.LayoutParams lp = new LayoutParams(awardIconSize, awardIconSize);

		for (String a : mAvards) {
			final ImageView img = new ImageView(getContext());
			vg.addView(img);
			img.setLayoutParams(lp);

			final String url = a;
			img.post(new Runnable() {
				@Override
				public void run() {
					aq.id(img).image(url, true, true, 0, 0, null, R.anim.award_anim);

				}
			});
		}
	}


}

File: app/src/main/java/ds/photosight/ui/widget/VotesWidget.kt
package ds.photosight.ui.widget

import android.animation.ValueAnimator
import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.view.animation.DecelerateInterpolator
import android.widget.ImageView
import android.widget.LinearLayout
import androidx.annotation.DrawableRes
import androidx.core.view.isVisible
import androidx.core.view.updatePadding
import ds.photosight.R
import ds.photosight.utils.dp
import kotlinx.android.synthetic.main.votes.view.*

class VotesWidget(context: Context, attrs: AttributeSet) : LinearLayout(context, attrs) {

    companion object {
        private const val ANIMATION_DURATION = 2000L
    }

    data class Stats(
        val views: Int = 0,
        val art: Int = 0,
        val original: Int = 0,
        val tech: Int = 0,
        val likes: Int = 0,
        val dislikes: Int = 0
    ) {
        var rating: Float = 1 / (likes + dislikes).toFloat() * likes.toFloat()
    }

    private var stats: Stats = Stats()

    @DrawableRes
    private var awardsResIds: List<Int> = emptyList()

    init {
        View.inflate(context, R.layout.votes, this)
    }

    fun init(stats: Stats, @DrawableRes awards: List<Int>) {
        this.stats = stats
        this.awardsResIds = awards
        awardsContainer.isVisible = false
        awardsContainer.removeAllViews()
    }

    fun runAnimations() {
        labelsContainer.startLayoutAnimation()

        ValueAnimator.ofFloat(0f, 1f)
            .setDuration(ANIMATION_DURATION)
            .apply {
                interpolator = DecelerateInterpolator(1.5f)
                addUpdateListener {
                    val value = it.animatedValue as Float
                    updateView(value)

                }
            }
            .start()

        if (awardsResIds.isNotEmpty()) addAwards()

    }

    private fun updateView(factor: Float) {
        fun Int.asText(factor: Float) = (this * factor).toInt().toString()

        views.text = stats.views.asText(factor)
        aRate.text = stats.art.asText(factor)
        oRate.text = stats.original.asText(factor)
        tRate.text = stats.tech.asText(factor)
        lRate.text = stats.likes.asText(factor)
        dRate.text = stats.dislikes.asText(factor)

        val height = stats.rating * factor * redBar.height
        greenBar.layoutParams.height = height.toInt()
    }


    private fun addAwards() {
        for (id in awardsResIds) {
            val img = ImageView(context)
            img.setImageDrawable(context.getDrawable(id))
            img.updatePadding(bottom = 4.dp)
            awardsContainer.addView(img)
            //img.layoutParams = lp
        }
        awardsContainer.isVisible = true
    }


}
